src/rest/applications/mcm/resubmission.py

```diff
@@ -0,0 +1,428 @@
+    need to recreate all the requests using this new change and reinject them.
+
+    Attributes:
+        mcm: McM client instance
+    """
+
+    def __init__(self, mcm: McM) -> None:
+        self._mcm = mcm
+        self._invalidator = InvalidateDeleteRequests(mcm=self._mcm)
+        self.logger = self._mcm.logger
+
+    def _root_request_to_approve(self, root_prepid: str) -> RootRequestReset:
+        """
+        This is a customization for resetting a root request so
+        that it ends up in an 'approve/approved' status and could
+        be easily reused to recreate a chained request and inject it
+        again.
+
+        Args:
+            root_prepid: Root request's prepid
+
+        Returns:
+            An attribute describing how the root request was processed.
+        """
+        data: Union[dict, None] = self._mcm.get(
+            object_type="requests", object_id=root_prepid
+        )
+        if not data:
+            raise ValueError(f"Root request {root_prepid} not found")
+
+        approval = data.get("approval")
+        status = data.get("status")
+        if approval == "submit" and status == "done":
+            keep_output: Union[list[bool], None] = data.get("keep_output")
+            if not keep_output:
+                raise ValueError("Keep output is not defined")
+
+            if all(keep_output):
+                # Avoid resetting this kind of root request
+                # Just reuse it as it is
+                return RootRequestReset.KEEPS_OUTPUT
+            else:
+                self.logger.warning(
+                    (
+                        "Root request (%s) is already done, resetting it, "
+                        "announcing the invalidation, and setting it to approve/approved"
+                    ),
+                    root_prepid,
+                )
+                validation: Union[dict, None] = data.get("validation")
+                self.logger.debug("Validation results: %s", validation)
+                if not validation:
+                    raise ValueError("Validation results not available")
+
+                # Reset the request.
+                reset_result = self._mcm.reset(prepid=root_prepid)
+                if not reset_result:
+                    msg = f"Unable to reset the root request: {root_prepid}"
+                    self.logger.error(msg)
+                    raise RuntimeError(msg)
+
+                # Announce the invalidation.
+                self._invalidator._process_invalidation(request_prepids=[root_prepid])
+
+                # Force its status to approve/approved, re-including the validation.
+                data: dict = self._mcm.get(
+                    object_type="requests", object_id=root_prepid
+                )
+                assert isinstance(data, dict)
+
+                data["validation"] = validation
+                data["approval"] = "approve"
+                data["status"] = "approved"
```

Comment:
Is this happily bypassing the validation? If the root campaign was changed a new validation might be necessary...

Reply:
Yes, it is reusing the old validation results. My understanding is it is not strictly required to repeat it as, for example: [PDMVDEV-106](https://its.cern.ch/jira/projects/PDMVDEV/issues/PDMVDEV-106).